File: src/whisper_input/recorder.py
"""Audio recording with PyAudio."""

import logging
import threading
from typing import Optional

import numpy as np
import pyaudio

from .config import AudioConfig, RecordingConfig

logger = logging.getLogger(__name__)


class AudioRecorder:
    """Record audio from microphone using PyAudio.

    Keeps the audio stream open to minimize latency on recording start.
    """

    def __init__(self, audio_config: AudioConfig, recording_config: RecordingConfig):
        self.audio_config = audio_config
        self.recording_config = recording_config
        self.audio: Optional[pyaudio.PyAudio] = None
        self.stream: Optional[pyaudio.Stream] = None
        self.buffer: list[bytes] = []
        self.is_recording = False
        self.lock = threading.Lock()
        self.max_frames = int(
            recording_config.max_duration
            * audio_config.sample_rate
            / audio_config.chunk_size
        )

    def _audio_callback(self, in_data, frame_count, time_info, status):
        """PyAudio callback - runs in separate thread."""
        if status:
            logger.warning(f"PyAudio status: {status}")

        with self.lock:
            if self.is_recording:
                self.buffer.append(in_data)
                # Safety limit
                if len(self.buffer) >= self.max_frames:
                    logger.warning("Max recording duration reached")
                    self.is_recording = False

        return (None, pyaudio.paContinue)

    def open_stream(self):
        """Open audio stream (call once at startup for low latency)."""
        if self.stream is not None:
            return  # Already open

        if self.audio is None:
            self.audio = pyaudio.PyAudio()

        device_index = self.audio_config.device_index
        if device_index is not None:
            logger.info(f"Using audio device index: {device_index}")

        self.stream = self.audio.open(
            format=pyaudio.paInt16,
            channels=self.audio_config.channels,
            rate=self.audio_config.sample_rate,
            input=True,
            input_device_index=device_index,
            frames_per_buffer=self.audio_config.chunk_size,
            stream_callback=self._audio_callback,
            start=False,  # Don't start immediately
        )
        logger.info("Audio stream opened (ready for low-latency recording)")

    def start(self):
        """Start recording audio."""
        if self.stream is None:
            self.open_stream()

        with self.lock:
            self.buffer = []
            self.is_recording = True

        if not self.stream.is_active():
            self.stream.start_stream()

        logger.info("Recording started")

    def stop(self) -> bytes:
        """Stop recording and return audio data."""
        with self.lock:
            self.is_recording = False
            audio_data = b"".join(self.buffer)
            self.buffer = []

        # Don't close the stream - just stop it for reuse
        if self.stream and self.stream.is_active():
            self.stream.stop_stream()

        logger.info(f"Recording stopped, captured {len(audio_data)} bytes")
        return audio_data

    def get_audio_as_numpy(self, audio_data: bytes) -> np.ndarray:
        """Convert raw audio bytes to numpy array for Whisper."""
        audio_array = np.frombuffer(audio_data, dtype=np.int16)
        # Normalize to [-1, 1] range as float32
        return audio_array.astype(np.float32) / 32768.0

    def is_silent(self, audio_data: bytes, threshold: float = 0.01) -> bool:
        """Check if audio is mostly silence based on RMS energy.

        Args:
            audio_data: Raw audio bytes
            threshold: RMS threshold below which audio is considered silent
                       (0.01 works well for typical microphone input)

        Returns:
            True if audio is silent, False if speech detected
        """
        audio = self.get_audio_as_numpy(audio_data)
        rms = np.sqrt(np.mean(audio ** 2))
        logger.debug(f"Audio RMS energy: {rms:.4f} (threshold: {threshold})")
        return rms < threshold

    def list_devices(self) -> list[dict]:
        """List available audio input devices."""
        if self.audio is None:
            self.audio = pyaudio.PyAudio()
        devices = []
        for i in range(self.audio.get_device_count()):
            info = self.audio.get_device_info_by_index(i)
            if info["maxInputChannels"] > 0:
                devices.append(
                    {
                        "index": i,
                        "name": info["name"],
                        "channels": info["maxInputChannels"],
                        "sample_rate": int(info["defaultSampleRate"]),
                    }
                )
        return devices

    def close(self):
        """Clean up PyAudio resources."""
        if self.stream:
            if self.stream.is_active():
                self.stream.stop_stream()
            self.stream.close()
            self.stream = None
        if self.audio:
            self.audio.terminate()
            self.audio = None

File: whisper-typer-rs/src/recorder.rs
//! Audio recording with cpal.
//!
//! Keeps the audio stream open for low-latency recording start.
//! Captures 16kHz mono f32 audio suitable for Whisper.

use cpal::traits::{DeviceTrait, HostTrait, StreamTrait};
use cpal::{SampleRate, Stream, StreamConfig};
use std::sync::{Arc, Mutex};
use std::time::Instant;
use tracing::{debug, info, warn};

use crate::config::{AudioConfig, RecordingConfig, SilenceConfig};

/// Audio recorder with always-open stream for low-latency start.
pub struct AudioRecorder {
    config: AudioConfig,
    #[allow(dead_code)]
    recording_config: RecordingConfig,
    #[allow(dead_code)]
    silence_config: SilenceConfig,
    /// Shared state between the audio callback thread and the main thread.
    shared: Arc<SharedState>,
    /// The cpal stream handle. Kept alive to maintain the always-open stream.
    _stream: Option<Stream>,
}

struct SharedState {
    inner: Mutex<RecorderInner>,
}

struct RecorderInner {
    is_recording: bool,
    buffer: Vec<f32>,
    max_samples: usize,
    // Silence detection state
    silence_start: Option<Instant>,
    recording_start: Option<Instant>,
    should_auto_stop: bool,
}

impl AudioRecorder {
    pub fn new(
        audio_config: AudioConfig,
        recording_config: RecordingConfig,
        silence_config: SilenceConfig,
    ) -> Self {
        let max_samples =
            (recording_config.max_duration * audio_config.sample_rate as f64) as usize;

        let shared = Arc::new(SharedState {
            inner: Mutex::new(RecorderInner {
                is_recording: false,
                buffer: Vec::with_capacity(max_samples),
                max_samples,
                silence_start: None,
                recording_start: None,
                should_auto_stop: false,
            }),
        });

        Self {
            config: audio_config,
            recording_config,
            silence_config,
            shared,
            _stream: None,
        }
    }

    /// Open the audio stream. Call once at startup.
    pub fn open_stream(&mut self) -> Result<(), String> {
        if self._stream.is_some() {
            return Ok(());
        }

        let host = cpal::default_host();
        let device = host
            .default_input_device()
            .ok_or("No input audio device available")?;

        info!(
            "Using audio device: {}",
            device.name().unwrap_or("unknown".into())
        );

        let stream_config = StreamConfig {
            channels: self.config.channels,
            sample_rate: SampleRate(self.config.sample_rate),
            buffer_size: cpal::BufferSize::Fixed(self.config.chunk_size),
        };

        let shared = Arc::clone(&self.shared);
        let silence_threshold = self.silence_config.threshold;
        let silence_duration = self.silence_config.duration;
        let min_speech_duration = self.silence_config.min_speech_duration;
        let max_recording_duration = self.silence_config.max_recording_duration;

        let stream = device
            .build_input_stream(
                &stream_config,
                move |data: &[f32], _info: &cpal::InputCallbackInfo| {
                    let mut inner = shared.inner.lock().unwrap();

                    if !inner.is_recording {
                        return;
                    }

                    // Append samples to buffer
                    let remaining = inner.max_samples.saturating_sub(inner.buffer.len());
                    let to_copy = data.len().min(remaining);
                    inner.buffer.extend_from_slice(&data[..to_copy]);

                    if inner.buffer.len() >= inner.max_samples {
                        warn!("Max recording duration reached");
                        inner.is_recording = false;
                        inner.should_auto_stop = true;
                        return;
                    }

                    // Silence detection
                    if let Some(rec_start) = inner.recording_start {
                        let elapsed = rec_start.elapsed().as_secs_f64();

                        // Check max recording duration
                        if elapsed >= max_recording_duration {
                            info!("Max recording duration reached ({max_recording_duration}s)");
                            inner.should_auto_stop = true;
                            return;
                        }

                        // Don't check silence until minimum speech duration
                        if elapsed < min_speech_duration {
                            return;
                        }

                        // RMS energy
                        let rms = rms_energy(data);
                        let is_silent = rms < silence_threshold;

                        if is_silent {
                            let silence_start =
                                inner.silence_start.get_or_insert_with(Instant::now);
                            if silence_start.elapsed().as_secs_f64() >= silence_duration {
                                debug!("Silence detected for {silence_duration}s — auto-stopping");
                                inner.should_auto_stop = true;
                            }
                        } else {
                            inner.silence_start = None;
                        }
                    }
                },
                move |err| {
                    warn!("Audio stream error: {err}");
                },
                None, // timeout
            )
            .map_err(|e| format!("Failed to build input stream: {e}"))?;

        stream.play().map_err(|e| format!("Failed to start audio stream: {e}"))?;
        info!("Audio stream opened (ready for low-latency recording)");

        self._stream = Some(stream);
        Ok(())
    }

    /// Start recording audio.
    pub fn start(&self) {
        let mut inner = self.shared.inner.lock().unwrap();
        inner.buffer.clear();
        inner.is_recording = true;
        inner.silence_start = None;
        inner.recording_start = Some(Instant::now());
        inner.should_auto_stop = false;
        info!("Recording started");
    }

    /// Stop recording and return captured audio samples (f32, mono, 16kHz).
    pub fn stop(&self) -> Vec<f32> {
        let mut inner = self.shared.inner.lock().unwrap();
        inner.is_recording = false;
        let samples = std::mem::take(&mut inner.buffer);
        let duration = samples.len() as f64 / self.config.sample_rate as f64;
        info!("Recording stopped: {:.1}s ({} samples)", duration, samples.len());
        samples
    }

    /// Check if auto-stop was triggered by silence detection.
    pub fn should_auto_stop(&self) -> bool {
        self.shared.inner.lock().unwrap().should_auto_stop
    }

    /// Check if currently recording.
    #[allow(dead_code)]
    pub fn is_recording(&self) -> bool {
        self.shared.inner.lock().unwrap().is_recording
    }

    /// Check if audio data is silent.
    pub fn is_silent(samples: &[f32], threshold: f32) -> bool {
        let rms = rms_energy(samples);
        debug!("Audio RMS energy: {rms:.4} (threshold: {threshold})");
        rms < threshold
    }

    /// Get the sample rate.
    pub fn sample_rate(&self) -> u32 {
        self.config.sample_rate
    }
}

/// Calculate RMS energy of audio samples.
fn rms_energy(samples: &[f32]) -> f32 {
    if samples.is_empty() {
        return 0.0;
    }
    let sum_sq: f32 = samples.iter().map(|s| s * s).sum();
    (sum_sq / samples.len() as f32).sqrt()
}
